src/v/kafka/server/quota_manager.cc

```diff
@@ -36,54 +61,56 @@ quota_manager::quota_manager()
       config::shard_local_cfg().kafka_client_group_byte_rate_quota.bind())
   , _target_fetch_tp_rate_per_client_group(
       config::shard_local_cfg().kafka_client_group_fetch_byte_rate_quota.bind())
+  , _client_quotas{client_quotas}
   , _gc_freq(config::shard_local_cfg().quota_manager_gc_sec())
   , _max_delay(config::shard_local_cfg().max_kafka_throttle_delay_ms.bind()) {
-    _gc_timer.set_callback([this] {
-        auto full_window = _default_num_windows() * _default_window_width();
-        gc(full_window);
-    });
+    _gc_timer.set_callback([this]() { gc(); });
 }
 
 quota_manager::~quota_manager() { _gc_timer.cancel(); }
 
 ss::future<> quota_manager::stop() {
     _gc_timer.cancel();
-    return ss::make_ready_future<>();
+    co_await _gate.close();
 }
 
 ss::future<> quota_manager::start() {
-    _gc_timer.arm_periodic(_gc_freq);
-    return ss::make_ready_future<>();
+    if (ss::this_shard_id() == _client_quotas.shard_id()) {
+        co_await _client_quotas.reset(client_quotas_map_t{});
+        _gc_timer.arm_periodic(_gc_freq);
+    }
 }
 
-ss::future<quota_manager::client_quotas_t::iterator>
+ss::future<quota_manager::client_quotas_map_t::iterator>
 quota_manager::maybe_add_and_retrieve_quota(
   std::optional<std::string_view> quota_id, clock::time_point now) {
     // requests without a client id are grouped into an anonymous group that
     // shares a default quota. the anonymous group is keyed on empty string.
-    auto qid = quota_id ? *quota_id : "";
-
-    // find or create the throughput tracker for this client
-    //
-    // c++20: heterogeneous lookup for unordered_map can avoid creation of
-    // an sstring here but std::unordered_map::find isn't using an
-    // equal_to<> overload. this is a general issue we'll be looking at. for
-    // now, these client-name strings are small. This will be solved in
-    // c++20 via Hash::transparent_key_equal.
-    auto [it, inserted] = _client_quotas.try_emplace(
-      ss::sstring(qid),
-      client_quota{
-        now,
-        clock::duration(0),
-        {static_cast<size_t>(_default_num_windows()), _default_window_width()},
-        {static_cast<size_t>(_default_num_windows()), _default_window_width()},
-        /// pm_rate is only non-nullopt on the qm home shard
-        (ss::this_shard_id() == quota_manager_shard)
-          ? std::optional<token_bucket_rate_tracker>(
-            {*_target_partition_mutation_quota(),
-             static_cast<uint32_t>(_default_num_windows()),
-             _default_window_width()})
-          : std::optional<token_bucket_rate_tracker>()});
+    auto qid = quota_id.value_or("");
+
+    vassert(_client_quotas, "_client_quotas should have been initialized");
+
+    auto it = _client_quotas->find(qid);
+    auto inserted = false;
+    if (it == _client_quotas->end()) {
+        co_await container().invoke_on(
+          _client_quotas.shard_id(),
+          [qid, now](quota_manager& me) { return me.add_quota_id(qid, now); });
```

Comment:
I think you also need to hold a gate on both sides (the calling shard and the called shard), so perhaps it's worth moving the dispatch and gates holding into `add_quota_id` and avoiding the assert.

Reply:
The `_gate` is only used to fail the `gc()` fibre once `quota_manager.stop()` has been called, so I don't think we need to hold gates on the request path. On the request path, we can rely on the assumption that `quota_manager` is only stopped/destructed after any outstanding requests have been processed/failed:
https://github.com/redpanda-data/redpanda/blob/a7cd2b0de8fd494bf0c49292b658900416efff22/src/v/redpanda/application.cc#L321-L338